### ISSUE TITLE
Fork & Code Licensing question

### ISSUE THREAD

**Issue by saibotk** [tslater2006/esp32_flash_loader#6]
Hey :wave: 

I found this amazing project during my research and had to add some adjustments to support ESP32-C3 chips.

While doing so I refactored some parts and added some improvements through digging a bit deeper into the Application Image format.

Check out my up-to-date fork here: https://github.com/saibotk/ghidra-esp32-flash-loader

--

Sorry, I didn't mean to just do a shameless self-advertisement. 

I want to properly license the project to use a valid open-source license. Ideally Apache-2, as that is what Espressif uses for their stuff and what Java projects in general use often.

Because this project currently has no license file it has all rights reserved.
Thus, I wanted to ask if you can license this project or if you grant me permission to license your code contributions as Apache-2.

Since my fork also builds upon other forks, I want to ask all of you too for permission to license the code and your contributions under Apache-2.

@Ebiroll 
@austinc3030
@SpecialMike

Please let me know if you agree or have any comments or ideas :)

Sorry for the ping, and thank you all for your work!

**Comment by austinc3030** [comment]
Hey! Sorry for taking so long to reply. TBH I'm just now seeing this as I didn't know how to get notifications in github working.

I am a big proponent of open source and by my understanding of the various licenses feel like GPL3 best promotes the spirit of opensource. However, given that ghidra and Espressif both use Apache2, I think this makes sense to license it under Apache2.

I cannot speak for @Ebiroll, @SpecialMike, or @tslater2006, but as far as my contributions, I am fine with Apache2. I will be updating my repository with an Apache2 license file. 

Thanks for reaching out.

**Comment by tslater2006** [comment]
Hey @saibotk , sorry I didn't see this sooner. Only just now saw it when austinc3030 tagged me. I would be perfectly fine with Apache 2 for this. As long as @Ebiroll is OK with this license change I'll go ahead and set an official license in this repository. Thank you for taking open source licensing seriously and reaching out.

**Comment by tslater2006** [comment]
just a note: It looks like I would have to remove the SVD-Loader python script from the repository first, since that is actually licensed GPL3 (and would require that all code in here be licensed GPL.

**Comment by SpecialMike** [comment]
I'm fine with you doing anything with my contributions in https://github.com/SpecialMike/esp32_flash_loader, I just used it for something else I was working on at the time and won't work on it anymore.

**Comment by austinc3030** [comment]
> Hey [@saibotk](https://github.com/saibotk) , sorry I didn't see this sooner. Only just now saw it when austinc3030 tagged me. I would be perfectly fine with Apache 2 for this. As long as [@Ebiroll](https://github.com/Ebiroll) is OK with this license change I'll go ahead and set an official license in this repository. Thank you for taking open source licensing seriously and reaching out.

I agree its nice to see open source licensing being taken seriously. This was a very welcome ping!

> just a note: It looks like I would have to remove the SVD-Loader python script from the repository first, since that is actually licensed GPL3 (and would require that all code in here be licensed GPL.

So after researching this, I don't know if that is accurate. It looks like you can still license the repo under Apache2. Apache2 and GPL3 are considered "compatible licenses" where you can license the repo as Apache2 but still preserve the GPL3 license for that file. It just needs to be explicitly stated which license applies to which files.

Further, given that you are the original author (minus the loader script), you currently hold "All Rights Reserved". So I think you can change/"set" the license since you are the parent repo. Then its a matter of flowing the license down.

I think we can do that by submitting a PR to Ebiroll's repository to earmark the license change. Then me and specialmike can pull the license into our forks and preserve the commit history.

So I don't think its appropriate for me to implement a license in my repo, but rather wait until tslater2006 implements the license then flow it down to mine. 

Thoughts?

**Comment by saibotk** [comment]
Apache 2 sounds good and fits all other licenses in this ecosystem.

Glad to hear back 👍 

Please let me know when you added the license. And yes you can relicense your repository without asking all contributors. Thats agreed upon in GitHubs ToS.

After all of you licensed your changes, i can go ahead and add the Apache License to my fork.

It's nice to see such a sweet response! Licensing such stuff just makes it way easier for some people to use this code and just clarifies everything. 

I strongly believe in open source and contributing back changes.

**Comment by Ebiroll** [comment]
Yes, sorry for not getting back.  Apache 2 sounds good. Always fun when old work is picked up by someone.
Just tell me what to do. 

> I strongly believe in open source and contributing back changes.
Great. That is how xtensa support got merged into Ghidra,

Here is some more info:
I started using this instead, https://github.com/tenable/esp32_image_parser
I wrote an article about it here, https://medium.com/@olof-astrand/reverse-engineering-of-esp32-flash-dumps-with-ghidra-or-ida-pro-8c7c58871e68
Espressif also released official SVD files. https://github.com/espressif/svd
Here is origianal SVD loader source:
https://github.com/leveldown-security/SVD-Loader-Ghidra

**Comment by tslater2006** [comment]
@Ebiroll @saibotk @austinc3030 @SpecialMike  I've added the license file to this repository. Thank you all for the quick responses :)

[fb9f4c4](https://github.com/tslater2006/esp32_flash_loader/commit/fb9f4c466731770861d683ec8fa36e685c8c94aa)

**Comment by austinc3030** [comment]
My repo has been updated with the license and it looks like ebiroll's has as well

**Comment by saibotk** [comment]
Thank you all very much! 

Heart-warming, that we coordinated this change and could finally add a proper license.

Added it to my repo too!
